caffe2/operators/spatial_batch_norm_op.h

```diff
@@ -90,35 +90,21 @@ class SpatialBNOp : public Operator<Context> {
           alpha_data,
           beta_data);
     } else {
-      auto* saved_mean = Output(SAVED_MEAN);
-      auto* saved_rstd = Output(SAVED_INV_STD);
-      if (num_batches_ == 1) {
-        saved_mean->Resize(C);
-        saved_rstd->Resize(C);
-      } else {
-        const auto& batch_mean_sum = Input(BATCH_MEAN_SUM);
-        const auto& batch_var_sum = Input(BATCH_VAR_SUM);
-        if (saved_mean != &batch_mean_sum) {
-          saved_mean->Resize(C);
-        }
-        if (saved_rstd != &batch_var_sum) {
-          saved_rstd->Resize(C);
-        }
-      }
+      auto* saved_mean = Output(SAVED_MEAN, {C}, at::dtype<T>());
+      auto* saved_rstd = Output(SAVED_INV_STD, {C}, at::dtype<T>());
       T* saved_mean_data = saved_mean->template mutable_data<T>();
       T* saved_rstd_data = saved_rstd->template mutable_data<T>();
-      auto* running_mean = Output(RUNNING_MEAN);
-      auto* running_var = Output(RUNNING_VAR);
-      if (running_mean->numel() != C) {
-        running_mean->Resize(C);
-        math::Set<T, Context>(
```

Comment:
Shouldn't you keep the `math::Set` calls, or are they done inside `Output`?

Reply:
this is checking whether running_mean has size C to decide whether the Tensor is being initialized or not(i.e. whether this is the first run), but I think the Tensor should be filled in the init_net? I remember I have discussed with dima before about this. cc @dzhulgakov @BIT-silence please take a look. I'm not sure whether we could refactor the function like this

Reply:
So ideally, we shouldn't rely on magic on-the-fly initialization as it can change the size of parameter/buffer. Thus ideally we'd add code here that enforces the size of the output to be correct. Note, that your current refactoring doesn't do that - if the output has the wrong size, we'd silently resize it and not initialize it at all. You can add an explicit ENFORCE on shape for the corresponding Input (as you already rely it being the same blob).

However, I'm a bit worried that there might be existing code relying on this behavior somewhere. The main wrapper (https://github.com/pytorch/pytorch/blob/master/caffe2/python/helpers/normalization.py#L129) does handle things correctly, but maybe there are other places.

To be safe - we can bring back the comparison / initialization code and add LOG_EVERY_MS() log to inform users that it's a broken behavior. You can check the size of input first as it's guaranteed that the output is the same.

Reply:
then how do we remove partially initialized Tensor?

Reply:
do you know what situation is this code trying to deal with? if the comparison is just for first run, we can probably add some boolean to mark whether this is the first run or not?

Reply:
yes, it's comparison for the first run effectively together with zero initialization of a tensor (and it's a bad design)

The hack you can do is check corresponding Input().size() (as it's "enforce in place") and if it's wrong - reset the output and do explicit zero-fill. It's if you want to play safe.

Another option is to just assert that the output already has the right size. It might be fine as most of the code I can find correctly initializes the buffers in init_net, so it would be fine.

Note, that it's not fully safe to do your current code - if the size is wrong we'd silently resize the output without zero-fill and continue populating the garbage memory

Reply:
going with the first option.